hudi-common/src/main/java/org/apache/hudi/common/table/read/HoodiePositionBasedFileGroupRecordBuffer.java

```diff
@@ -174,20 +189,97 @@ public boolean containsLogRecord(String recordKey) {
   }
 
   @Override
-  protected boolean doHasNext() throws IOException {
-    ValidationUtils.checkState(baseFileIterator != null, "Base file iterator has not been set yet");
-
-    // Handle merging.
-    while (baseFileIterator.hasNext()) {
-      T baseRecord = baseFileIterator.next();
-      nextRecordPosition = readerContext.extractRecordPosition(baseRecord, readerSchema, ROW_INDEX_TEMPORARY_COLUMN_NAME, nextRecordPosition);
-      Pair<Option<T>, Map<String, Object>> logRecordInfo = records.remove(nextRecordPosition++);
-      if (hasNextBaseRecord(baseRecord, logRecordInfo)) {
-        return true;
+  protected boolean hasNextBaseRecord(T baseRecord) throws IOException {
+    if (!readerContext.getShouldMergeUseRecordPosition()) {
+      return doHasNextFallbackBaseRecord(baseRecord);
+    }
+
+    nextRecordPosition = readerContext.extractRecordPosition(baseRecord, readerSchema,
+        ROW_INDEX_COLUMN_NAME, nextRecordPosition);
+    Pair<Option<T>, Map<String, Object>> logRecordInfo = records.remove(nextRecordPosition++);
+
+    Map<String, Object> metadata = readerContext.generateMetadataForRecord(
+        baseRecord, readerSchema);
+
+    Option<T> resultRecord = logRecordInfo != null
+        ? merge(Option.of(baseRecord), metadata, logRecordInfo.getLeft(), logRecordInfo.getRight())
+        : merge(Option.empty(), Collections.emptyMap(), Option.of(baseRecord), metadata);
+    if (resultRecord.isPresent()) {
+      nextRecord = readerContext.seal(resultRecord.get());
+      return true;
+    }
+    return false;
+  }
+
+  private boolean doHasNextFallbackBaseRecord(T baseRecord) throws IOException {
+    if (needToDoHybridStrategy) {
+      //see if there is a delete block with record positions
+      nextRecordPosition = readerContext.extractRecordPosition(baseRecord, readerSchema,
+          ROW_INDEX_TEMPORARY_COLUMN_NAME, nextRecordPosition);
+      Pair<Option<T>, Map<String, Object>> logRecordInfo  = records.remove(nextRecordPosition++);
```

Comment:
Does it need to be thread-safe? Or can multiple threads update the nextRecordPosition? Maybe make it AtomicLong?

Reply:
My understanding is that FG reader is single threaded